File: WebMap.Web/src/types/static-marker.ts
import type { Vector2 } from '@/types/math'
import type { SeString } from '@/types/string'

export const StaticMarkerType = {
    Normal: 0,
    Title: 1,
} as const
export type StaticMarkerType = typeof StaticMarkerType[keyof typeof StaticMarkerType]

export const StaticMarkerDataType = {
    Normal: 0,
    MapLink: 1,
    ImplicitMapLink: 2,
    Aetheryte: 3,
    PlaceName: 4,
} as const
export type StaticMarkerDataType = typeof StaticMarkerDataType[keyof typeof StaticMarkerDataType]

export const SubtextOrientation = {
    Left: 1,
    Right: 2,
    Bottom: 3,
    Top: 4,
} as const
export type SubtextOrientation = typeof SubtextOrientation[keyof typeof SubtextOrientation] | number

interface StaticMarkerBase {
    title?: SeString
    icon?: number
    position: Vector2
    subtextOrientation: SubtextOrientation | number
    type: StaticMarkerType
    dataType: number
}

export interface NormalStaticMarker extends StaticMarkerBase {
    dataType: typeof StaticMarkerDataType.Normal
}

export interface MapLinkStaticMarker extends StaticMarkerBase {
    dataType: typeof StaticMarkerDataType.MapLink
    targetMapId: number
}

export interface ImplicitMapLinkStaticMarker extends StaticMarkerBase {
    dataType: typeof StaticMarkerDataType.ImplicitMapLink
    targetMapId: number
}

export interface AetheryteStaticMarker extends StaticMarkerBase {
    dataType: typeof StaticMarkerDataType.Aetheryte
    aetheryteId: number
}

export interface PlaceNameStaticMarker extends StaticMarkerBase {
    dataType: typeof StaticMarkerDataType.PlaceName
    tooltip: SeString
}

export interface UnknownStaticMarker extends StaticMarkerBase {
    dataKey: number
}

export type StaticMarker =
    | NormalStaticMarker
    | MapLinkStaticMarker
    | ImplicitMapLinkStaticMarker
    | AetheryteStaticMarker
    | PlaceNameStaticMarker
    | UnknownStaticMarker

File: WebMap.Plugin/Models/StaticMarker.cs
using System;
using System.Numerics;

namespace WebMap.Models;

public abstract class StaticMarker
{
    public required ReadOnlyMemory<byte>? Title { get; init; }
    public required ushort? Icon { get; init; }
    public required Vector2 Position  { get; init; }
    public required byte SubtextOrientation { get; init; }
    public required StaticMarkerType Type { get; init; }
    
    // ReSharper disable once UnusedMember.Global
    public abstract StaticMarkerDataType DataType { get; }
}

public sealed class NormalStaticMarker : StaticMarker
{
    public override StaticMarkerDataType DataType => StaticMarkerDataType.Normal;
}

public sealed class MapLinkStaticMarker : StaticMarker
{
    public override StaticMarkerDataType DataType => StaticMarkerDataType.MapLink;
    public required uint TargetMapId { get; init; }
}

public sealed class ImplicitMapLinkStaticMarker : StaticMarker
{
    public override StaticMarkerDataType DataType => StaticMarkerDataType.ImplicitMapLink;
    public required uint TargetMapId { get; init; }
}

public sealed class AetheryteStaticMarker : StaticMarker
{
    public override StaticMarkerDataType DataType => StaticMarkerDataType.Aetheryte;
    public required uint AetheryteId { get; init; }
}

public sealed class PlaceNameStaticMarker : StaticMarker
{
    public override StaticMarkerDataType DataType => StaticMarkerDataType.PlaceName;
    public required ReadOnlyMemory<byte> Tooltip { get; init; }
}

public sealed class UnknownStaticMarker(byte dataType) : StaticMarker
{
    public override StaticMarkerDataType DataType => new(dataType);
    public required uint DataKey { get; init; }
}

public readonly struct StaticMarkerType(byte value) : IRawValue<byte>
{
    public byte RawValue { get; } = value;
    
    public const byte NormalValue = 0;
    public const byte TitleValue = 1;
    
    public static readonly StaticMarkerType Normal = new(NormalValue);
    public static readonly StaticMarkerType Title = new(TitleValue);
}

public readonly struct StaticMarkerDataType(byte value) : IRawValue<byte>
{
    public byte RawValue { get; } = value;
    
    public const byte NormalValue = 0;
    public const byte MapLinkValue = 1;
    public const byte ImplicitMapLinkValue = 2;
    public const byte AetheryteValue = 3;
    public const byte PlaceNameValue = 4;

    public static readonly StaticMarkerDataType Normal = new(NormalValue);
    public static readonly StaticMarkerDataType MapLink = new(MapLinkValue);
    public static readonly StaticMarkerDataType ImplicitMapLink = new(ImplicitMapLinkValue);
    public static readonly StaticMarkerDataType Aetheryte = new(AetheryteValue);
    public static readonly StaticMarkerDataType PlaceName = new(PlaceNameValue);
}
